### PR TITLE
Virtual pump bug863

### PR DESCRIPTION
**Pull request by andyrozman**
Implementation of Virtual pumps selection. Issue #863

### CODE REVIEW

**Comment by andyrozman** [comment]
The classes were intended to be used by anybody. You could for example delete the whole pumpDescription block from *any* pump plugin implementation (which is why I put them in PumpCommon. 
For example for combo: 
setPumpDescription(pumpDescription, PumpType.AccuChekCombo); 
and you can loose whole initialization block. 

It would be more practical if each of Plugin Type, would have it's own parent so you would put all Pump stuff into info.nightscout.androidaps.plugins.Pumps. Since we don't have common parent, I decided to put them in PumpCommon, where we could put stuff that could be used by all Pump implementations.

**Comment by MilosKozak** [comment]
05-10 10:18:28.529 9680-9680/info.nightscout.androidaps E/AndroidRuntime: FATAL EXCEPTION: main
    Process: info.nightscout.androidaps, PID: 9680
    java.lang.RuntimeException: Unable to resume activity {info.nightscout.androidaps/info.nightscout.androidaps.MainActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean info.nightscout.androidaps.plugins.PumpCommon.defs.PumpCapability.hasCapability(info.nightscout.androidaps.plugins.PumpCommon.defs.PumpCapability)' on a null object reference
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3103)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3134)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1388)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:148)
        at android.app.ActivityThread.main(ActivityThread.java:5417)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean info.nightscout.androidaps.plugins.PumpCommon.defs.PumpCapability.hasCapability(info.nightscout.androidaps.plugins.PumpCommon.defs.PumpCapability)' on a null object reference
        at info.nightscout.androidaps.plugins.PumpCommon.utils.PumpUtil.setPumpDescription(PumpUtil.java:45)
        at info.nightscout.androidaps.plugins.PumpVirtual.VirtualPumpPlugin.refreshConfiguration(VirtualPumpPlugin.java:429)
        at info.nightscout.androidaps.plugins.PumpVirtual.VirtualPumpFragment$2.run(VirtualPumpFragment.java:108)
        at android.app.Activity.runOnUiThread(Activity.java:5511)
        at info.nightscout.androidaps.plugins.PumpVirtual.VirtualPumpFragment.updateGUI(VirtualPumpFragment.java:88)
        at info.nightscout.androidaps.plugins.Common.SubscriberFragment.onResume(SubscriberFragment.java:21)
        at android.support.v4.app.Fragment.performResume(Fragment.java:2387)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1467)
        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1752)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1821)
        at android.support.v4.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3251)
        at android.support.v4.app.FragmentManagerImpl.dispatchResume(FragmentManager.java:3215)
        at android.support.v4.app.FragmentController.dispatchResume(FragmentController.java:217)
        at android.support.v4.app.FragmentActivity.onResumeFragments(FragmentActivity.java:509)
        at android.support.v4.app.FragmentActivity.onPostResume(FragmentActivity.java:498)
        at android.support.v7.app.AppCompatActivity.onPostResume(AppCompatActivity.java:171)
        at android.app.Activity.performResume(Activity.java:6336)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3092)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3134) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1388) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:148) 
        at android.app.ActivityThread.main(ActivityThread.java:5417) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)

**Comment by MilosKozak** [comment]
05-10 10:24:56.092 13499-13499/info.nightscout.androidaps E/AndroidRuntime: FATAL EXCEPTION: main
    Process: info.nightscout.androidaps, PID: 13499
    java.lang.RuntimeException: Unable to resume activity {info.nightscout.androidaps/info.nightscout.androidaps.MainActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean info.nightscout.androidaps.plugins.PumpCommon.defs.PumpCapability.hasCapability(info.nightscout.androidaps.plugins.PumpCommon.defs.PumpCapability)' on a null object reference
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3103)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3134)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1388)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:148)
        at android.app.ActivityThread.main(ActivityThread.java:5417)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean info.nightscout.androidaps.plugins.PumpCommon.defs.PumpCapability.hasCapability(info.nightscout.androidaps.plugins.PumpCommon.defs.PumpCapability)' on a null object reference
        at info.nightscout.androidaps.plugins.PumpCommon.utils.PumpUtil.setPumpDescription(PumpUtil.java:45)
        at info.nightscout.androidaps.plugins.PumpVirtual.VirtualPumpPlugin.refreshConfiguration(VirtualPumpPlugin.java:429)
        at info.nightscout.androidaps.plugins.PumpVirtual.VirtualPumpFragment$2.run(VirtualPumpFragment.java:108)
        at android.app.Activity.runOnUiThread(Activity.java:5511)
        at info.nightscout.androidaps.plugins.PumpVirtual.VirtualPumpFragment.updateGUI(VirtualPumpFragment.java:88)
        at info.nightscout.androidaps.plugins.Common.SubscriberFragment.onResume(SubscriberFragment.java:21)
        at android.support.v4.app.Fragment.performResume(Fragment.java:2387)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1467)
        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1752)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1821)
        at android.support.v4.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3251)
        at android.support.v4.app.FragmentManagerImpl.dispatchResume(FragmentManager.java:3215)
        at android.support.v4.app.FragmentController.dispatchResume(FragmentController.java:217)
        at android.support.v4.app.FragmentActivity.onResumeFragments(FragmentActivity.java:509)
        at android.support.v4.app.FragmentActivity.onPostResume(FragmentActivity.java:498)
        at android.support.v7.app.AppCompatActivity.onPostResume(AppCompatActivity.java:171)
        at android.app.Activity.performResume(Activity.java:6336)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3092)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3134) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1388) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:148) 
        at android.app.ActivityThread.main(ActivityThread.java:5417) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)

**Comment by andyrozman** [comment]
I fixed the bug yesterday and the commits were added to my branch. Ready to be tested again.

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -1085,6 +1085,10 @@
     <string name="openapsama_max_daily_safety_multiplier">Max daily safety multiplier</string>
     <string name="openapsama_current_basal_safety_multiplier">Current basal safety multiplier</string>
     <string name="value_unavailable_short">n/a</string>
+    <string name="virtualpump_type">Virtual Pump Type</string>
+    <string name="virtualpump_definition">Pump Definition</string>
+    <string name="virtualpump_pump_def" formated="false">Bolus: Step=%s\nExtended Bolus: [Step=%s, Duration=%smin-%sh]\nBasal: Step=%s\nTBR: %s (by %s), Duration=%smin-%sh\nEXTENDED_NOTE</string>
```
Comment (MilosKozak) [inline app/src/main/res/values/strings.xml]: use positional formating here pls

File: app/src/main/java/info/nightscout/androidaps/plugins/PumpVirtual/VirtualPumpFragment.java [hunk 2]
```diff
@@ -97,6 +104,20 @@ public void run() {
...
+
+                    virtualPump.refreshConfiguration();
+
+                    PumpType pumpType = virtualPump.getPumpType();
+
+                    pumpTypeView.setText(pumpType.getDescription());
+
+                    String template = MainApp.gs(R.string.virtualpump_pump_def);
+
+                    template = template.replace("EXTENDED_NOTE", pumpType.hasExtendedBasals() ? //
```
Comment (MilosKozak) [inline app/src/main/java/info/nightscout/androidaps/plugins/PumpVirtual/VirtualPumpFragment.java]: EXTENDED_NOTE is really ugly hack and it's hard to translate. move the note string to pump definitions where needed

**Comment by MilosKozak** [comment]
@andyrozman fix these 2 comments and add new logging method with `L.PUMP` like here
https://github.com/MilosKozak/AndroidAPS/blob/dev/app/src/main/java/info/nightscout/androidaps/plugins/PumpDanaRS/DanaRSPlugin.java#L68
and here
https://github.com/MilosKozak/AndroidAPS/blob/dev/app/src/main/java/info/nightscout/androidaps/plugins/PumpDanaRS/DanaRSPlugin.java#L186

It's good 👍 I'll merge it to 2.0

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -1087,7 +1087,7 @@
     <string name="value_unavailable_short">n/a</string>
     <string name="virtualpump_type">Virtual Pump Type</string>
     <string name="virtualpump_definition">Pump Definition</string>
-    <string name="virtualpump_pump_def" formated="false">Bolus: Step=%s\nExtended Bolus: [Step=%s, Duration=%smin-%sh]\nBasal: Step=%s\nTBR: %s (by %s), Duration=%smin-%sh\nEXTENDED_NOTE</string>
+    <string name="virtualpump_pump_def" formated="false">Bolus: Step=%1$s\nExtended Bolus: [Step=%2$s, Duration=%3$smin-%4$sh]\nBasal: Step=%5$s\nTBR: %6$s (by %7$s), Duration=%8$smin-%9$sh\n%10$s</string>
```
[the remark this answers is not in the document]
Comment (andyrozman) [reply app/src/main/res/values/strings.xml]: Fixed. I also changed PumpType (removed one of Medtronic pumps) and changed arrays.xml.

**Comment by MilosKozak** [comment]
fixed some bugs and merged.
THANKS!

**Pull request merged**